pravega/pravega
# Issue 2804: Client reply with server stack trace on error

**Change log description**
This PR allows Pravega to send server-side stack traces to clients upon failed requests to ease debugging activities:
- Pravega servers and controller instances may now attach internal server-side stack traces to clients on failed requests.
- Due to security reasons, this behavior is activated via configuration switches in both components.

**Purpose of the change**
Fixes #2804.

**What the code does**
This code enables Pravega Servers and Controller instances to send stack traces to the client as part of the reply on a failed request.

On the one hand, `PravegaRequestProcessor` and `AppendProcessor` in the SegmentStore are now capable of sending server traces to the client, given that error messages in `WireCommands` now include a new serializable field `serverStackTrace` to encapsulate that information. In both cases, the server-side stack trace is sent to the client if we set `pravegaservice.replyWithStackTraceOnError=true` in the `config.properties` file.

On the other hand, this PR includes changes in `ControllerServiceImpl` to allow the Controller to reply clients with server-side traces on GRPC calls. To activate this behavior, we need to set `REPLY_WITH_STACK_TRACE_ON_ERROR = true` in the `reference.conf` file.

**How to verify it**
All tests should be working as before this PR. Moreover, when activating the `replyWithStackTraceOnError` option, we should see new traces in client logs. In the following, there is an example of the effects of activating the `replyWithStackTraceOnError` option in the test driver log of system tests.

_Without SegmentStore client traces (i.e., standard behavior):_
```
2018-09-10 10:10:10,497 122959 [epollEventLoopGroup-5-4] INFO  i.p.c.s.i.SegmentOutputStreamImpl - Received SegmentSealed WireCommands.SegmentIsSealed(type=SEGMENT_IS_SEALED, requestId=7068, segment=testAutoScale16234781/testScaleUp/0.#epoch.0, serverStackTrace=) on writer 1069d20d-6d8d-427e-a0be-fcd677ad844e
```

_With SegmentStore client traces:_
```
2018-09-10 09:42:41,939 123922 [epollEventLoopGroup-5-2] INFO  i.p.c.s.i.SegmentOutputStreamImpl - Received SegmentSealed WireCommands.SegmentIsSealed(type=SEGMENT_IS_SEALED, requestId=7810, segment=testAutoScale1415837769/testScaleUp/0.#epoch.0, serverStackTrace=io.pravega.segmentstore.contracts.StreamSegmentSealedException: [Segment 'testAutoScale1415837769/testScaleUp/0.#epoch.0'] The StreamSegment is sealed and cannot be modified.
    at io.pravega.segmentstore.server.logs.SegmentMetadataUpdateTransaction.preProcessOperation(SegmentMetadataUpdateTransaction.java:228)
    at io.pravega.segmentstore.server.logs.ContainerMetadataUpdateTransaction.preProcessOperation(ContainerMetadataUpdateTransaction.java:312)
    at io.pravega.segmentstore.server.logs.OperationMetadataUpdater.preProcessOperation(OperationMetadataUpdater.java:224)
    at io.pravega.segmentstore.server.logs.OperationProcessor.processOperation(OperationProcessor.java:364)
    at io.pravega.segmentstore.server.logs.OperationProcessor.processOperations(OperationProcessor.java:283)
    at java.util.concurrent.CompletableFuture.uniAccept(CompletableFuture.java:656)
    at java.util.concurrent.CompletableFuture$UniAccept.tryFire(CompletableFuture.java:632)
    at java.util.concurrent.CompletableFuture$Completion.run(CompletableFuture.java:442)
    at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
    at java.util.concurrent.FutureTask.run(FutureTask.java:266)
    at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180)
    at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293)
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
    at java.lang.Thread.run(Thread.java:745)
) on writer fe64b810-dac7-4774-a3c7-c9fcfe2f16a3
```
